### DISCUSSION TITLE
Mac-style Keyboard Bindings on a Windows Keyboard with Alt as Cmd

### DISCUSSION THREAD

**Discussion by ubunatic** [sezanzeb/input-remapper#608] [Show and tell]
First of all, Tobi thanks for this great tool! @sezanzeb
I have tried different things over the years and having a working GUI + start and stop button makes a huge difference than fiddling with clunky config files. 🙇🏼 

I have used input-mapper now for a few days and at first thought I could not get the mapping I needed. I was aiming for the following with Alt working like the MacOs Cmd key.

* Alt + Tab: toggle current and previous app on single press
* Alt + [Tab|Shift+Tab ...]: switch app in the app switcher (needs holding Alt)
* Alt + (c|v|x): Copy paste
* Alt + (q|d|w|t|...): Quit, bookmark, close tab, open tab, etc. (most common Cmd bindings)
* Alt + v: paste in the terminal (no need to hold Shift when pasting)
* Alt + c: do not send ^C to the terminal (using Ctrl + Insert instead)
* Alt (+ Shift) + [Left|Right]: as Ctrl + (Left|Right) for word navigation
* Ctrl + Alt + Arrows: as Super + Arrows to move Windows in Wayland/Gnome
* Alt (+ Shift) + Click: as Ctrl + Click for multi selection in sheets and file browser
* Super_L: as AltGr for easy typing of @, €, \, |, ~, etc. on a German layout
* Capslock: as Super (for window manager things)
* Alt: single Alt must be ignored (to avoid app menus popping up)

Quite a list and there are event some more things I needed while using my bindings.
But I think I got it actually working and have put my first version of a binding file here:
https://codeberg.org/ubunatic/dotfiles/src/branch/main/config/input-mapper-mac-bindings.json

It still has some bugs, such as the Alt key being stuck sometimes, resulting in Ctrl being pressed and not coming up. This does not happen often though.

One of the fancy things that worked was mapping Alt to Control and still keep both needed Alt + Tab bindings bound to Alt + Tab.

1. Map Alt to hold Ctrl and also set a virtual `$alt` down, then release both and also release Alt.
```
# Alt_L
key_down(Control_L)
.set(alt, Down).hold()
.set(alt, Up).key_up(Alt_L)
.key_up(Control_L)
```

2. Map Alt + Tab to Alt + Tab, but ignore the Ctrl key and use virtual `$alt` to keep Alt_L pressed.
```
# Alt_L + Tab
if_eq($alt, Down,
  key_up(Control_L).
  key_down(Alt_L).k(Tab),
  modify(Alt_L, k(Tab))
)
```

Not sure if it this over engineered. At least it works and all other attempts failed.
I had trouble with the Alt key being not held down and the app switcher not opening.
Also single Alt presses popped up the app menus, which you do not want when Alt becomes such a central command key.

3. To paste in the terminal with Alt + v, I use Shift + Insert and also needed to ignore Ctrl.
```
# Alt_L + v
key_up(Control_L)
.hold_keys(Shift_L, Insert)
.hold()
.key_down(Control_L)
```

Most other bindings are straight forward and I added anything that was missing when my muscle memory was looking for it.

If I update my bindings, I will also post an update here.

**Let me know if or when there is a better approach to realize these bindings.**
```
> input-remapper-control -v
input-remapper 1.6.0-beta 9adc300aeefe39c94655738ed52e848808f991f3 https://github.com/sezanzeb/input-remapper
python-evdev 1.6.1
```

PS: I am a Mac Keyboarder for a year now and really prefer the Cmd over the Ctrl key for copy/paste and most commands. Using the thumb (on Alt) feels much better then the pinky or thumb (on Ctrl). I enter most commands one-handed now.

**Comment by jonasBoss** [comment]
Glad to see the beta branch getting some love.

I think the else arm in the `Alt_L + Tab` macro will not do what you want: it will release the alt key if you release either the tab or the alt key.
and you don't need any variables for this, because the kernel keeps track of all the pressed keys.
```python
# Alt L:
hold(Control_L)
.key_up(Alt_L) # alt may or may not be down. The kernel ignores duplicate events

# Alt L + Tab
key_up(Control_L)  # ctrl is down if we pressed alt first, if not the kernel will ignore this anyway
.key_down(Alt_L) # the `Alt L` macro will call key_up as soon as `Alt L` is released
.hold(Tab)
```

**Reply by ubunatic** [reply]
Yes I realized this while playing further with the options.
I will update my config after some cleanup.

**Comment by HibGut** [comment]
Thanks for this! I was looking for a solution for Mac-style key bindings. But your link to your bindings file is wrong. There's no useful content there.

**Reply by mathiasrando** [reply]
You can find the file contents by using the history:
https://codeberg.org/ubunatic/dotfiles/src/commit/9c359ecabe93928521bbaa3f7774ad6b7fd19a07/config/Mac%20Bindings.json
